src/treemodel/composer/modifyselection.js

```diff
@@ -27,20 +27,23 @@ export default function modifySelection( selection, options = {} ) {
 	const focus = selection.focus;
 	const walker = new TreeWalker( {
 		boundaries: getSearchRange( focus, isForward ),
-		singleCharacters: true
+		singleCharacters: true,
+		direction: isForward ? 'FORWARD' : 'BACKWARD'
 	} );
 
-	const items = Array.from( walker );
-	let next = items[ isForward ? 'shift' : 'pop' ]();
+	let next = walker.next();
 
 	// 1. Nothing to do here.
-	if ( !next ) {
+	if ( next.done ) {
 		return;
 	}
 
+	// Replace TreeWalker step wrapper by clean step value.
+	next = next.value;
```

Comment:
It was today morning (before coffee) idea :) https://github.com/ckeditor/ckeditor5-engine/pull/425/commits/d7c1f7d1cebadb56b8d8826f376563ed27c38924